buildengine/project.go

```diff
@@ -0,0 +1,153 @@
+package buildengine
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"golang.design/x/reflect"
+
+	"github.com/TBD54566975/ftl/common/moduleconfig"
+)
+
+// Project models FTL modules and external libraries and is used to manage dependencies within the build engine
+type Project interface {
+	Config() ProjectConfig
+	CopyWithDependencies([]ProjectKey) Project
+	String() string
+}
+
+type ProjectConfig struct {
+	Key          ProjectKey
+	Dir          string
+	Language     string
+	Watch        []string
+	Dependencies []ProjectKey
+}
+
+var _ = (Project)(Module{})
+var _ = (Project)(ExternalLibrary{})
+
+// Module represents an FTL module in the build engine
+type Module struct {
+	moduleconfig.ModuleConfig
+	Dependencies []string
+}
+
+func (m Module) Config() ProjectConfig {
+	return ProjectConfig{
+		Key:          ProjectKey(m.ModuleConfig.Module),
+		Dir:          m.ModuleConfig.Dir,
+		Language:     m.ModuleConfig.Language,
+		Watch:        m.ModuleConfig.Watch,
+		Dependencies: ProjectKeysFromModuleNames(m.Dependencies),
+	}
+}
+
+func (m Module) CopyWithDependencies(dependencies []ProjectKey) Project {
+	module := reflect.DeepCopy(m)
+	module.Dependencies = StringsFromProjectKeys(dependencies)
+	return Project(module)
+}
+
+func (m Module) String() string {
+	return "module " + m.ModuleConfig.Module
+}
+
+// ExternalLibrary represents a library that makes use of FTL modules, but is not itself an FTL module
+type ExternalLibrary struct {
+	Dir          string
+	Language     string
+	Dependencies []string
+}
+
+func (e ExternalLibrary) Config() ProjectConfig {
+	var watch []string
+	switch e.Language {
+	case "go":
+		watch = []string{"**/*.go", "go.mod", "go.sum"}
+	case "kotlin":
+		watch = []string{"pom.xml", "src/**", "target/generated-sources"}
+	default:
+		panic(fmt.Sprintf("unknown language %T", e.Language))
+	}
+
+	return ProjectConfig{
+		Key:          ProjectKey("lib:" + e.Dir),
+		Dir:          e.Dir,
+		Language:     e.Language,
+		Watch:        watch,
+		Dependencies: ProjectKeysFromModuleNames(e.Dependencies),
+	}
+}
+
+func (e ExternalLibrary) CopyWithDependencies(dependencies []ProjectKey) Project {
+	lib := reflect.DeepCopy(e)
+	lib.Dependencies = StringsFromProjectKeys(dependencies)
+	return Project(lib)
+}
+
+func (e ExternalLibrary) String() string {
+	return "library " + e.Dir
+}
+
+// ProjectKey is a unique identifier for the project (ie: a module name or a library path)
+// It is used to:
+// - build the dependency graph
+// - map changes in the file system to the project
+type ProjectKey string
+
+func ProjectKeyForModuleName(name string) ProjectKey {
```

Comment:
This function is pretty redundant given that a cast is shorter and is identical in effect.